src/responsibleai/rai_analyse/generate_pdf.py

```diff
@@ -0,0 +1,1226 @@
+    cohort_performance_section = cohort_performance_section + "".join(
+        cohort_performance_containers["error_analysis_min"])
+
+    return str(heading_section + cohort_performance_section)
+
+
+def get_causal_page(data):
+    left_elem = [div(p("Causal analysis answers real world what if "
+                       "questions about how changes of treatments would impact a real world outcome."))]
+
+    left_container = div(left_elem, _class="left")
+
+    main_elems = []
+
+    def get_causal_dot_plot(center, em, ep):
+        png_base64 = get_dot_plot(center, em, ep)
+        return div(img(_src="data:image/png;base64,{}".format(png_base64)), _class="image_div")
+
+    def get_table_row(data):
+        table_row_elems = []
+        for v in data:
+            table_row_elems.append(td(v, _class="cell"))
+        return tr(table_row_elems, _class="row")
+
+    def get_table(data):
+        horizontal_headings = ["Index", "Current<br>Treatment", "Recommended<br>Treatment", "Effect<br>Estimate"]
+        headings_td = [td(x, _class="header_cell") for x in horizontal_headings]
+        headings = thead(tr(headings_td, _class="row"), _class="table-head")
+
+        rows_elems = []
+        for elem in data:
+            rows_elems.append(get_table_row(elem))
+
+        body = tbody(rows_elems, _class="table-body")
+
+        return table(headings, body, _class="table")
+
+    for f in data["global_effect"].values():
+        main_elems.append(
+            div(
+                h3(f["feature"]),
+                p(
+                    "On average, increasing \"{}\" by 1 unit increases the prediction outcome by {}".format(
+                        f["feature"], round(f["point"], 3)
+                    )
+                ),
+                get_causal_dot_plot(f["point"], f["ci_upper"] - f["point"], f["point"] - f["ci_lower"]),
+                _class="nobreak_div"
+            )
+        )
+
+        main_elems.append(
+            h3(
+                "Top data points responding the most to treatment on \"{}\":".format(f["feature"])
+            )
+        )
+        main_elems.append(
+            p(
+                "What data points experience the largest impact (causal effect) to changes"
+                ", while treating \"{}\"".format(f["feature"])
+            )
+        )
+
+        main_elems.append(
+            get_table(
+                [[p["index"], p["Current treatment"], p["Treatment"], round(p["Effect of treatment"], 2)]
+                 for p in data["top_local_policies"][f["feature"]]])
+        )
+
+    main_container = div(main_elems, _class="main")
+
+    return div(
+        div(
+            get_page_divider("Causal"),
+            left_container,
+            main_container,
+            _class="nobreak_div"
+        ), _class="container nobreak_div")
+
+
+class ClassificationComponents:
+    @staticmethod
+    def get_model_overview_page(data):
+        return get_model_overview(data)
```

Comment:
This one doesn't look recursive, but..... why?

Reply:
the *components class shares the same composition for the pages creation method for the workflow class in create_rai_insight. This is attempting to reuse components common for both classification and regression.

resolved with generate_pdf.py refactoring into individual components.